libkbfs/block_journal.go

```diff
@@ -978,179 +945,85 @@ func (j *blockJournal) ignoreBlocksAndMDRevMarkers(ctx context.Context,
 }
 
-// onMDFlush removes at most `maxToRemove` blocks from the
-// `saveUntilMDFlush` journal if one exists.  If `lastToRemove` is
-// zero, it flushes the complete `saveUntilMDFlush` if it has fewer
-// than `maxToRemove` entries; if it doesn't flush the entire journal,
-// it returns the ordinal of the current last entry.  If
-// `lastToRemove` is non-zero, it only flushes up to the minimum of
-// `lastToRemove` and the earliest entry + `maxToRemove`; if this
-// flushes the entire saved journal or reaches a non-ignored marker
-// for MD revision `flushedMDRev`, it returns 0, otherwise it returns
-// `lastToRemove`.  It's intended that the caller should call this
-// function repeatedly until it returns 0, releasing any locks in
-// between calls so it doesn't block other operations for too long.
-func (j *blockJournal) onMDFlush(ctx context.Context,
-	maxToRemove uint64, flushedMDRev MetadataRevision,
-	lastToRemove journalOrdinal) (
-	nextLastToRemove journalOrdinal,
-	removedBytes, removedFiles int64, err error) {
-	if j.saveUntilMDFlush == nil {
-		return 0, 0, 0, nil
-	}
-
-	if maxToRemove == 0 {
-		return 0, 0, 0, errors.New("maxToRemove must be non-zero")
-	}
-
-	// Delete the block data for anything in the saved journal.
-	first, err := j.saveUntilMDFlush.readEarliestOrdinal()
+// getDeferredRange gets the earliest and latest revision of the
+// deferred GC journal.  If the returned length is 0, there's no need
+// for further GC.  Should be called under lock.
+func (j *blockJournal) getDeferredGCRange() (
+	len int, earliest, latest journalOrdinal, err error) {
+	earliest, err = j.deferredGC.readEarliestOrdinal()
 	if ioutil.IsNotExist(err) {
 		return 0, 0, 0, nil
 	} else if err != nil {
 		return 0, 0, 0, err
 	}
 
-	last, err := j.saveUntilMDFlush.readLatestOrdinal()
+	latest, err = j.deferredGC.readLatestOrdinal()
 	if ioutil.IsNotExist(err) {
 		return 0, 0, 0, nil
 	} else if err != nil {
 		return 0, 0, 0, err
 	}
 
-	if lastToRemove != 0 {
-		if last < lastToRemove {
-			return 0, 0, 0, errors.Errorf("Last removal requested is %d, but "+
-				"last entry in journal is %d", lastToRemove, last)
-		} else if first > lastToRemove {
-			return 0, 0, 0, errors.Errorf("Last removal requested is %d, but first "+
-				"entry in journal is %d", lastToRemove, first)
-		}
-
-		last = lastToRemove
-	}
-
-	lastMin := last
-	if max := journalOrdinal(maxToRemove); lastMin > first+max-1 {
-		j.log.CDebugf(ctx, "Last removal requested is %d, but capping at %d",
-			last, first+max-1)
-		lastMin = first + max - 1
-	}
-
-	j.log.CDebugf(ctx, "Removing saved data for entries [%d, %d]",
-		first, lastMin)
-	for i := first; i <= lastMin; i++ {
-		e, err := j.saveUntilMDFlush.readJournalEntry(i)
-		if err != nil {
-			return 0, 0, 0, err
-		}
+	return int(latest - earliest + 1), earliest, latest, nil
+}
 
-		_, err = j.saveUntilMDFlush.removeEarliest()
+// onMDFlush drains the deferred GC journal.  This doesn't touch the
```

Comment:
Maybe this comment could be rewritten to avoid explicitly mentioning locks. Something like:

```
doGC collects any unreferenced blocks from flushed entries. earliest and latest should be
from a call to getDeferredGCRange, and clearDeferredGCRange should be called after
this function. This function only reads the deferred GC journal at the given range and
reads/writes the block store, so callers may use that to relax any synchronization requirements.
```